neps/nep-0518.md

```diff
@@ -0,0 +1,221 @@
+
+Note: some Ethereum transaction fields are intentionally omitted because they are unused by the wallet contract.
+
+These fields are used to validate the transaction and derive Near actions that the wallet contract will create as receipts. The details of this process are described below.
+
+##### Ethereum transaction validation
+
+The following validation conditions MUST pass for the wallet contract to accept a transaction.
+
+1. `from` address when formatted as hex-encoded with `0x` prefix MUST match the current account ID (i.e. the wallet contract's account ID).
+2. `chain_id` MUST match one of the following values depending on the Near chain the wallet contract is deployed to: mainnet -> 397; testnet -> 398; any other chain -> 399. The mainnet and testnet values are registered with the [official Ethereum ecosystem registry of chain IDs](https://github.com/ethereum-lists/chains).
+3. `nonce` MUST match the nonce value currently stored in the contract state.
+4. `to` address MUST either (a) be equal to `keccak256(target)[12,32]` (where `target` is other argument passed to the `rlp_execute` function) or (b) when `to` is formatted as hex-encoded with `0x` prefix it MUST be equal to `target`. In case (b) there is an additional validation check that the `to` address is not registered in the "Ethereum Translation Contract" (ETC). The details of this check and why it is needed are discussed in Appendix C.
+5. `value` MUST be less than or equal to `(2**128 - 1) // 1_000_000`. This condition arises from the mismatch is decimal places between Ether and NEAR which is discussed in the definition of **Wei** in Appendix A. Essentially, we must ensure the `value` can be mapped into a valid amount of yoctoNEAR, which means `value * 1_000_000 <= u128::MAX`.
+
+##### Converting Ethereum transaction into Near actions
+
+Each Ethereum transaction is converted to a single Near action (batch transactions are not supported) based on the `data` field. Following the Solidity convention of the first four bytes of the data being a **method selector**, the wallet contract checks the first four bytes of the `data` to see if it is a known Near action. The **method selectors** for Near actions supported by the wallet contract are determined by mapping the actions to an equivalent Solidity function signature as follows:
+
+- `functionCall(string,string,bytes,uint64,uint32)`
+- `transfer(string,uint32)`
+- `addKey(uint8,bytes,uint64,bool,bool,uint128,string,string[])`
+- `deleteKey(uint8,bytes)`
+
+Note that the `uint32` fields in `functionCall` and `transfer` contain the amount of yoctoNEAR tha cannot be included in the Ethereum transaction's `value` field due to the difference in decimal places (see **Wei** definition in Appendix A), therefore the value there is always less than `1_000_000` so it will easily fit in a 32-bit number. These type signatures then hash to the **method selectors**:
```

Comment:
```suggestion
Note that the `uint32` fields in `functionCall` and `transfer` contain the amount of yoctoNEAR that cannot be included in the Ethereum transaction's `value` field due to the difference in decimal places (see **Wei** definition in Appendix A), therefore the value there is always less than `1_000_000` so it will easily fit in a 32-bit number. These type signatures then hash to the **method selectors**:
```

Reply:
Done in https://github.com/near/NEPs/pull/555/commits/1f64d1fec30ef7e70a6a0675eb459a3eae1a6315